File: lib/core/calculators/shrimp_respiration_calculator.dart
import 'dart:convert';
import 'dart:math'; // Import needed for clamp
import 'package:flutter/foundation.dart'; // For kIsWeb and potentially debugPrint
import 'package:flutter/services.dart' show rootBundle;

/// Calculates shrimp respiration rate based on loaded data using trilinear interpolation.
class ShrimpRespirationCalculator {
  final String dataPath; // Path to the JSON data file
  Map<String, dynamic>? _respirationData; // Parsed JSON data (nested map)
  // Lists to store the discrete values from the data grid for interpolation bounds
  List<double> _salinityValues = [];
  List<double> _temperatureValues = [];
  List<double> _biomassValues = [];

  ShrimpRespirationCalculator(this.dataPath);

  // --- Getters for Testing ---
  // These expose internal state for verification in unit tests.
  @visibleForTesting
  List<double> get salinityValuesForTest => List.unmodifiable(_salinityValues);
  @visibleForTesting
  List<double> get temperatureValuesForTest => List.unmodifiable(_temperatureValues);
  @visibleForTesting
  List<double> get biomassValuesForTest => List.unmodifiable(_biomassValues);
  @visibleForTesting
  Map<String, dynamic>? get respirationDataForTest => _respirationData == null ? null : Map.unmodifiable(_respirationData!);
  // --- End Getters for Testing ---


  /// Loads and parses the respiration data from the JSON file specified by [dataPath].
  Future<void> loadData() async {
    try {
      final String jsonString = await rootBundle.loadString(dataPath);
      final Map<String, dynamic> jsonData = jsonDecode(jsonString);

      // Extract metadata to define the grid dimensions and values
      final metadata = jsonData['metadata'];
      if (metadata == null) throw Exception("Metadata missing in JSON");

      // Parse and store the discrete values for each dimension, ensuring they are sorted
      _salinityValues = (metadata['salinity_values'] as List)
          .map((s) => double.parse(s.toString().replaceAll('%', ''))) // Ensure parsing from string
          .toList()..sort(); // Sort for reliable boundary finding
      _temperatureValues = (metadata['temperature_values'] as List)
          .map((t) => double.parse(t.toString().replaceAll('°C', ''))) // Ensure parsing from string
          .toList()..sort();
      _biomassValues = (metadata['shrimp_biomass'] as List)
          .map((b) => double.parse(b.toString().replaceAll('g', ''))) // Ensure parsing from string
          .toList()..sort();

      // Store the main data grid
      _respirationData = jsonData['data'];
      if (_respirationData == null) throw Exception("Data grid missing in JSON");

      debugPrint("Shrimp respiration data loaded successfully."); // Use debugPrint

    } catch (e) {
      debugPrint("Error loading shrimp respiration data from $dataPath: $e"); // Use debugPrint
      throw Exception('Failed to load shrimp respiration data: $e');
    }
  }

  /// Calculates the respiration rate (mg O₂/g/h) for the given conditions.
  /// Uses trilinear interpolation based on the loaded data.
  double getRespirationRate(double salinity, double temperature, double shrimpWeight) {
    if (_respirationData == null || _salinityValues.isEmpty || _temperatureValues.isEmpty || _biomassValues.isEmpty) {
      throw Exception('Respiration data not loaded or invalid. Call loadData() first.');
    }

    // 1. Clamp input values to the range covered by the data
    final clampedSalinity = salinity.clamp(_salinityValues.first, _salinityValues.last);
    final clampedTemperature = temperature.clamp(_temperatureValues.first, _temperatureValues.last);
    final clampedWeight = shrimpWeight.clamp(_biomassValues.first, _biomassValues.last);

    // 2. Find the surrounding grid points (lower and upper bounds for each dimension)
    //    `lastWhere` finds the largest value <= input, `firstWhere` finds the smallest value >= input.
    final salinityLow = _salinityValues.lastWhere((s) => s <= clampedSalinity, orElse: () => _salinityValues.first);
    final salinityHigh = _salinityValues.firstWhere((s) => s >= clampedSalinity, orElse: () => _salinityValues.last);
    final tempLow = _temperatureValues.lastWhere((t) => t <= clampedTemperature, orElse: () => _temperatureValues.first);
    final tempHigh = _temperatureValues.firstWhere((t) => t >= clampedTemperature, orElse: () => _temperatureValues.last);
    final weightLow = _biomassValues.lastWhere((w) => w <= clampedWeight, orElse: () => _biomassValues.first);
    final weightHigh = _biomassValues.firstWhere((w) => w >= clampedWeight, orElse: () => _biomassValues.last);

    // 3. Convert boundary values to the string keys used in the JSON data
    final salinityLowKey = '${salinityLow.toInt()}%';
    final salinityHighKey = '${salinityHigh.toInt()}%';
    final tempLowKey = '${tempLow.toInt()}°C';
    final tempHighKey = '${tempHigh.toInt()}°C';
    final weightLowKey = '${weightLow.toInt()}g';
    final weightHighKey = '${weightHigh.toInt()}g';

    // 4. Retrieve the respiration rate values at the 8 corners of the interpolation cube.
    //    Assumes the JSON structure: data[salinity][temperature][weight]
    //    Includes null checks for safer access.
    try {
      // Use helper for safer access
      final r000 = _getValueFromData(salinityLowKey, tempLowKey, weightLowKey);
      final r001 = _getValueFromData(salinityLowKey, tempLowKey, weightHighKey);
      final r010 = _getValueFromData(salinityLowKey, tempHighKey, weightLowKey);
      final r011 = _getValueFromData(salinityLowKey, tempHighKey, weightHighKey);
      final r100 = _getValueFromData(salinityHighKey, tempLowKey, weightLowKey);
      final r101 = _getValueFromData(salinityHighKey, tempLowKey, weightHighKey);
      final r110 = _getValueFromData(salinityHighKey, tempHighKey, weightLowKey);
      final r111 = _getValueFromData(salinityHighKey, tempHighKey, weightHighKey);


      // Check if any corner value is null (indicates missing data in JSON)
      final cornerValues = [r000, r001, r010, r011, r100, r101, r110, r111];
      if (cornerValues.any((r) => r == null)) {
         throw Exception("Missing respiration data for interpolation corner points near S=$salinity, T=$temperature, W=$shrimpWeight. Check JSON structure and keys (e.g., $salinityLowKey, $tempLowKey, $weightLowKey).");
      }

      // Cast non-null values
      final double r000d = r000!;
      final double r001d = r001!;
      final double r010d = r010!;
      final double r011d = r011!;
      final double r100d = r100!;
      final double r101d = r101!;
      final double r110d = r110!;
      final double r111d = r111!;


      // 5. Calculate interpolation factors (relative position within the cube, 0 to 1)
      //    Denominator is clamped to avoid division by zero if bounds are the same.
      final double salDiff = (salinityHigh - salinityLow);
      final double tempDiff = (tempHigh - tempLow);
      final double weightDiff = (weightHigh - weightLow);

      // Avoid division by zero if bounds are identical
      final s = salDiff == 0 ? 0.0 : (clampedSalinity - salinityLow) / salDiff;
      final t = tempDiff == 0 ? 0.0 : (clampedTemperature - tempLow) / tempDiff;
      final w = weightDiff == 0 ? 0.0 : (clampedWeight - weightLow) / weightDiff;


      // 6. Perform trilinear interpolation:
      //    Interpolate along the weight axis (w) at each of the 4 temperature/salinity corners
      final r00 = r000d + (r001d - r000d) * w;
      final r01 = r010d + (r011d - r010d) * w;
      final r10 = r100d + (r101d - r100d) * w;
      final r11 = r110d + (r111d - r110d) * w;

      //    Interpolate along the temperature axis (t) using the results from the weight interpolation
      final r0 = r00 + (r01 - r00) * t;
      final r1 = r10 + (r11 - r10) * t;

      //    Interpolate along the salinity axis (s) using the results from the temperature interpolation
      final respirationRate = r0 + (r1 - r0) * s;

      return respirationRate;

    } catch (e) {
       // Catch potential errors during map lookup (e.g., key not found) or casting
       debugPrint("Error during respiration data lookup/interpolation: $e");
       throw Exception("Failed to calculate respiration rate. Check JSON data structure and keys near S=$salinity, T=$temperature, W=$shrimpWeight.");
    }
  }

  /// Safely retrieves and converts a value from the nested respiration data map.
  double? _getValueFromData(String salKey, String tempKey, String weightKey) {
      final level1 = _respirationData?[salKey];
      if (level1 is! Map) return null;
      final level2 = level1[tempKey];
      if (level2 is! Map) return null;
      final value = level2[weightKey];
      if (value is num) {
        return value.toDouble();
      }
      return null; // Return null if key path doesn't exist or value is not a number
  }
}

File: backend/shrimp_respiration_calculator.py
"""Shrimp respiration rate calculator with trilinear interpolation."""
import os
from typing import Any, Dict, List, Optional, Tuple

from utils import load_json_data


class ShrimpRespirationCalculator:
    """
    Calculates shrimp respiration rate based on loaded data using
    trilinear interpolation. Loads data from a JSON file structured
    similarly to the Dart version's input.
    """

    def __init__(self, data_path: Optional[str] = None):
        """
        Initializes the calculator with the path to the JSON data file.

        Args:
            data_path: The path to the JSON data file.
            If None, computed dynamically.
        """
        if data_path is None:
            script_dir = os.path.dirname(os.path.abspath(__file__))
            repo_root = os.path.dirname(script_dir)
            self.data_path = os.path.join(
                repo_root, "assets", "data",
                "shrimp_respiration_salinity_temperature_weight.json"
            )
        else:
            self.data_path = data_path

        if not os.path.exists(self.data_path):
            raise FileNotFoundError(
                f"Shrimp respiration data file not found at: {self.data_path}"
            )

        self._respiration_data: Dict[str, Any] = {}
        self._salinity_values: List[float] = []
        self._temperature_values: List[float] = []
        self._biomass_values: List[float] = []
        self.load_data()

    def load_data(self) -> None:
        """Loads and parses the respiration data from the JSON file."""
        print(
            f"Attempting to load shrimp respiration data from: "
            f"{self.data_path}"
        )
        json_data = load_json_data(self.data_path)

        metadata: Dict[str, Any] = json_data.get('metadata', {})
        if not metadata:
            raise ValueError("Metadata missing or invalid in JSON")

        self._salinity_values = sorted([
            float(str(s).replace('%', ''))
            for s in metadata.get('salinity_values', [])
        ])
        self._temperature_values = sorted([
            float(str(t).replace('°C', ''))
            for t in metadata.get('temperature_values', [])
        ])
        self._biomass_values = sorted([
            float(str(b).replace('g', ''))
            for b in metadata.get('shrimp_biomass', [])
        ])

        if not all([
            self._salinity_values,
            self._temperature_values,
            self._biomass_values
        ]):
            raise ValueError(
                "Metadata arrays (salinity, temperature, biomass) "
                "cannot be empty"
            )

        self._respiration_data = json_data.get('data', {})
        if not self._respiration_data:
            raise ValueError(
                "Data grid missing or invalid in JSON"
            )

        print("Shrimp respiration data loaded successfully.")

    def _find_bounds(
        self, value: float, sorted_values: List[float]
    ) -> Tuple[float, float]:
        """Finds the lower and upper bounds for a value in a sorted list."""
        if not sorted_values:
            raise ValueError("Cannot find bounds in empty list")

        low = sorted_values[0]
        high = sorted_values[-1]

        for val in sorted_values:
            if val <= value:
                low = val
            if val >= value:
                high = val
                break

        return low, high

    def _get_value_from_data(
        self, sal_key: str, temp_key: str, weight_key: str
    ) -> float:
        """
        Safely retrieves a value from the nested respiration data dictionary.
        """
        try:
            val = self._respiration_data.get(sal_key, {}).get(temp_key, {}) \
                .get(weight_key)
            if val is None:
                raise ValueError(
                    (
                        f"Key path not found: {sal_key} -> {temp_key} -> "
                        f"{weight_key}"
                    )
                )
            return float(val)
        except (TypeError, ValueError) as e:
            raise ValueError(
                f"Could not convert value at {sal_key}.{temp_key}."
                f"{weight_key}: {e}"
            ) from e

    def _get_cube_values(
        self,
        salinity_low_key: str,
        salinity_high_key: str,
        temp_low_key: str,
        temp_high_key: str,
        weight_low_key: str,
        weight_high_key: str
    ) -> List[float]:
        """Retrieves the 8 corner values for the interpolation cube."""
        corners = [
            (salinity_low_key, temp_low_key, weight_low_key),
            (salinity_low_key, temp_low_key, weight_high_key),
            (salinity_low_key, temp_high_key, weight_low_key),
            (salinity_low_key, temp_high_key, weight_high_key),
            (salinity_high_key, temp_low_key, weight_low_key),
            (salinity_high_key, temp_low_key, weight_high_key),
            (salinity_high_key, temp_high_key, weight_low_key),
            (salinity_high_key, temp_high_key, weight_high_key)
        ]
        values = [
            self._get_value_from_data(sal, temp, weight)
            for sal, temp, weight in corners
        ]
        return values

    def _trilinear_interpolation(
        self,
        corner_values: List[float],
        s: float,
        t: float,
        w: float
    ) -> float:
        """
        Performs trilinear interpolation using the corner values and
        interpolation factors.
        """
        r000, r001, r010, r011, r100, r101, r110, r111 = corner_values
        r00 = r000 + (r001 - r000) * w
        r01 = r010 + (r011 - r010) * w
        r10 = r100 + (r101 - r100) * w
        r11 = r110 + (r111 - r110) * w
        r0 = r00 + (r01 - r00) * t
        r1 = r10 + (r11 - r10) * t
        return r0 + (r1 - r0) * s

    def get_respiration_rate(
        self, salinity: float, temperature: float, shrimp_weight: float
    ) -> float:
        """
        Calculates the respiration rate (mg O₂/g/h) using trilinear
        interpolation.

        Args:
            salinity: Water salinity in ppt.
            temperature: Water temperature in °C.
            shrimp_weight: Average shrimp weight in grams.

        Returns:
            The estimated respiration rate in mg O₂/g/h.

        Raises:
            ValueError: If data is not loaded or interpolation fails.
        """
        if not all([
            self._salinity_values,
            self._temperature_values,
            self._biomass_values
        ]):
            raise ValueError(
                "Respiration data not loaded. Call load_data() first."
            )

        # Clamp input values
        clamped_salinity = max(
            self._salinity_values[0],
            min(salinity, self._salinity_values[-1])
        )
        clamped_temperature = max(
            self._temperature_values[0],
            min(temperature, self._temperature_values[-1])
        )
        clamped_weight = max(
            self._biomass_values[0],
            min(shrimp_weight, self._biomass_values[-1])
        )

        # Find bounds
        salinity_low, salinity_high = self._find_bounds(
            clamped_salinity, self._salinity_values
        )
        temp_low, temp_high = self._find_bounds(
            clamped_temperature, self._temperature_values
        )
        weight_low, weight_high = self._find_bounds(
            clamped_weight, self._biomass_values
        )

        # Convert to JSON keys
        salinity_low_key = f'{int(salinity_low)}%'
        salinity_high_key = f'{int(salinity_high)}%'
        temp_low_key = f'{int(temp_low)}°C'
        temp_high_key = f'{int(temp_high)}°C'
        weight_low_key = f'{int(weight_low)}g'
        weight_high_key = f'{int(weight_high)}g'

        # Get corner values
        corner_values = self._get_cube_values(
            salinity_low_key, salinity_high_key, temp_low_key, temp_high_key,
            weight_low_key, weight_high_key
        )

        # Calculate interpolation factors
        sal_diff = salinity_high - salinity_low
        temp_diff = temp_high - temp_low
        weight_diff = weight_high - weight_low
        s = 0.0 if sal_diff == 0 else (
            clamped_salinity - salinity_low
        ) / sal_diff
        t = 0.0 if temp_diff == 0 else (
            clamped_temperature - temp_low
        ) / temp_diff
        w = 0.0 if weight_diff == 0 else (
            clamped_weight - weight_low
        ) / weight_diff

        # Perform interpolation
        return self._trilinear_interpolation(corner_values, s, t, w)


if __name__ == '__main__':
    try:
        calculator = ShrimpRespirationCalculator()
        SAL1, TEMP1, WEIGHT1 = 20.0, 28.0, 12.0
        rate1 = calculator.get_respiration_rate(SAL1, TEMP1, WEIGHT1)
        print(
            f"Respiration Rate at S={SAL1}‰, T={TEMP1}°C, W={WEIGHT1}g: "
            f"{rate1:.4f} mg O₂/g/h"
        )

        SAL2, TEMP2, WEIGHT2 = 25.0, 30.0, 15.0
        rate2 = calculator.get_respiration_rate(SAL2, TEMP2, WEIGHT2)
        print(
            f"Respiration Rate at S={SAL2}‰, T={TEMP2}°C, W={WEIGHT2}g: "
            f"{rate2:.4f} mg O₂/g/h"
        )

        SAL3, TEMP3, WEIGHT3 = 50.0, 5.0, 1.0
        rate3 = calculator.get_respiration_rate(SAL3, TEMP3, WEIGHT3)
        print(
            f"Respiration Rate at S={SAL3}‰, T={TEMP3}°C, W={WEIGHT3}g "
            f"(clamped): {rate3:.4f} mg O₂/g/h"
        )
    except (FileNotFoundError, ValueError, TypeError) as exc:
        print(f"An error occurred: {exc}")
